src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -169,21 +197,73 @@ private void RegisterNotification(Action action)
 
             private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
             {
-                OnDiagnosticsUpdated(e);
+                lock (_taskGate)
+                {
+                    // Chain the events so we process them serially.  This also ensures
+                    // that we don't process events while still getting our initial set
+                    // of diagnostics.
+                    var asyncToken = _owner._listener.BeginAsyncOperation(GetType() + ".OnDiagnosticsUpdated");
+                    _taskChain = _taskChain.ContinueWith(
+                        _ => OnDiagnosticsUpdatedOnBackground(e), TaskScheduler.Default);
+                    _taskChain.CompletesAsyncOperation(asyncToken);
+                }
+            }
+
+            private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
+            {
+                this.AssertIsBackground();
+                RegisterNotification(() => OnDiagnosticsUpdatedOnForeground(e));
```

Comment:
why does this move to background thread and go back to foreground thread without doing anything in background thread?